### PR TITLE
RFC: Add code::blocks backend

### PR DESCRIPTION
**Pull request by madebr**
- The backend uses the _NinjaBackend_ for generating the build files
- I have tested it with c and cpp projects using the gcc compiler.

How likely is it this gets merged?
See mesonbuild/meson#1300

### CODE REVIEW

File: mesonbuild/backend/codeblocksbackend.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        ET.SubElement(project_el, 'Option', {'virtualFolders': ''})
+
+        build_el = ET.SubElement(project_el, 'Build')
+        target_all_el = ET.SubElement(build_el, 'Target', {'title': 'all'})
+        ET.SubElement(target_all_el, 'Option', {'working_dir': self.environment.build_dir})
+        ET.SubElement(target_all_el, 'Option', {'type': str(self.BUILD_OPTION_COMMANDS_ONLY)})
+
+        target_all_commands = ET.SubElement(target_all_el, 'MakeCommands')
+        ET.SubElement(target_all_commands, 'Build', {'command': '"{}" -v all'.format(self.ninja_bin)})
+        ET.SubElement(target_all_commands, 'CompileFile', {'command':'"{}" -v "$file"'.format(self.ninja_bin)})
```
Comment (ignatenkobrain) [inline mesonbuild/backend/codeblocksbackend.py]: __[flake8]__
* [E231] missing whitespace after ':'

File: mesonbuild/backend/codeblocksbackend.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        ET.SubElement(target_all_commands, 'Clean', {'command':'"{}" -v clean'.format(self.ninja_bin)})
```
Comment (ignatenkobrain) [inline mesonbuild/backend/codeblocksbackend.py]: __[flake8]__
* [E231] missing whitespace after ':'

File: mesonbuild/backend/codeblocksbackend.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        ET.SubElement(target_all_commands, 'DistClean', {'command':'"{}" -f "$makefile" -v clean'.format(self.ninja_bin)})
```
Comment (ignatenkobrain) [inline mesonbuild/backend/codeblocksbackend.py]: __[flake8]__
* [E231] missing whitespace after ':'

**Comment by jpakkane** [comment]
We don't want to add backends for all the different IDEs. Instead what we provide is the tooling and integration bits necessary so that IDEs can extract all the information they need to represent things in their internal representation. This is what Gnome Builder does and there is work underway (I'm told) to do the same with Qt Creator.

We have to generate XCode and VS project files out of necessity, because they do not support this but instead insist on their own project formats. Other IDEs should use `meson introspect` instead. This allows them to do their extra special features (such as debugging support and the like) without needing us. Also we don't have the manpower to maintain many IDE generators because they are fairly labour intensive.

**Comment by john-preston** [comment]
@jpakkane About the work for Qt Creator.

There is this issue in Qt Creator tracker: https://bugreports.qt.io/browse/QTCREATORBUG-18117

The problem described there states:

> I have been working on and off on meson support in the meantime, but unfortunately meson's suggested way to integrate into an IDE requires lots of meson processes to start. With an IDE doing the forking, that takes a very long time. I am talking to upstream, maybe we can find a more efficient way to get to the data.
> 
> Just as a reference point: I killed creator 3min after clicking on systemd's meson.build file. It was still busy gathering data about targets.

and they've lowered priority from "P2: Important" to "P4: Low" :(

Is this them doing something in a wrong way or do they have a point and integration API should be improved in some way?

**Comment by hunger** [comment]
I only glanced over this, but I think for Qt Creator the information provided in this backend is incomplete.
We used to use CBP files for our cmake support and the results were sub-optimal:-(

We need information on the toolchains being used, the compiler flags for all files, include paths, etc. Creator in effect builds each and every file -- so it needs pretty much all the information that a build system has.

I really appreciate that meson tried to stay out of the backend business as far as possible and think this patch is a step into the wrong direction. I would like Creator to be able to open *any* existing build directory of meson with all the expected configuration and things. The introspection API enables that, while this patch will lead to IDEs being able to only open a subset of build directories.

This patch goes into the same direction that made CMake's IDE support so terrible for years. They did now spend three years mopping up -- and they are still struggling. Please do not repeat their mistake!

**Pull request closed without merging**